File: 01.Stack/20.有效的括号.py
# -*- coding: utf-8 -*-
#
#
# [20] 有效的括号
#

class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if not s or len(s) == 0:
            return True

        if len(s) % 2 == 1:
            return False

        t = []

        for c in s:
            if c == '[' or c == '{' or c == '(':
                t.append(c)
            elif c == ']':
                last = t.pop() if t else '#'
                if last != '[':
                    return False
            elif c == '}':
                last = t.pop() if t else '#'
                if last != '{':
                    return False
            elif c == ')':
                last = t.pop() if t else '#'
                if last != '(':
                    return False
            else:
                return False

        return not t

File: 01.Stack/20.有效的括号.cpp
/*
 *
 * [20] 有效的括号
 */

class Solution {
   public:
    bool isValid(string s) {
        if (s.length() & 0x01) {
            return false;
        }
        stack<char> t;
        for (auto c : s) {
            if (c == '{' || c == '[' || c == '(') {
                t.push(c);
            } else if (c == ')') {
                if (t.empty() || t.top() != '(') {
                    return false;
                }
                t.pop();
            } else if (c == '}') {
                if (t.empty() || t.top() != '{') {
                    return false;
                }
                t.pop();
            } else if (c == ']') {
                if (t.empty() || t.top() != '[') {
                    return false;
                }
                t.pop();
            } else {
                return false;
            }
        }

        return t.empty();
    }
};
